src/Xrefcheck/Core.hs

```diff
@@ -71,66 +70,58 @@ instance Given ColorMode => Buildable Position where
 
 -- | Full info about a reference.
 data Reference = Reference
-  { rName   :: Text
+  { rName :: Text
     -- ^ Text displayed as reference.
-  , rLink   :: Text
-    -- ^ File or site reference points to.
-  , rAnchor :: Maybe Text
-    -- ^ Section or custom anchor tag.
-  , rPos    :: Position
+  , rPos :: Position
     -- ^ Position in source file.
-  , rInfo   :: ReferenceInfo
-    -- ^ More info about the link.
+  , rInfo :: ReferenceInfo
+    -- ^ More info about the reference.
   } deriving stock (Show, Generic)
 
 -- | Info about the reference.
 data ReferenceInfo
-  = RIExternal
-    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@
-  | RIOtherProtocol
-    -- ^ Entry not to be processed, e.g. @mailto:e-mail@
-  | RIFileLocal
-    -- ^ Reference to this file, e.g. @[a](#header)@
-  | RIFileAbsolute
-    -- ^ Reference to a file absolute to the root, e.g. @[c](/folder/file#header)@
-  | RIFileRelative
-    -- ^ Reference to a file relative to given one, e.g. @[b](folder/file#header)@
+  = RIExternal ExternalLink
+  | RIFile ReferenceInfoFile
+  deriving stock (Show, Generic)
+
+data ReferenceInfoFile = ReferenceInfoFile
+  { rifAnchor :: Maybe Text
+    -- ^ Section or custom anchor tag.
+  , rifLink :: FileLink
+    -- ^ More info about the link.
+  } deriving stock (Show, Generic)
+
+data ExternalLink
+  = ELUrl Text
+    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@.
+  | ELOther Text
+    -- ^ Entry not to be processed, e.g. @mailto:e-mail@.
+  deriving stock (Show, Generic)
+
+data FileLink
+  = FLAbsolute Text
+    -- ^ Reference to a file relative to the repository root.
+  | FLRelative Text
+    -- ^ Reference to a file relative to the given one.
```

Comment:
These two `Text` types could be changed also to `RelPosixLink`, but then the `Reference` `Buildable`  instance would show the link as normalised instead of as it is written in the Markdown file.

Would that be fine? Or it is preferable to keep the raw link here as it is.

Reply:
The original one, for displaying, would also be with POSIX path separators although not normalised? Otherwise, as the original one could be platform-dependant, the program output would also be platform-dependant.

Reply:
I found some trouble when trying to keep both the canonical and the original path because the resulting code was not so neat. I tried then a different approach: keep the original path and implement "normalised" `Eq` and `Ord` instances but reached a similar situation.

Finally, I have tried to define a `RelPosixLink` that is just a relative POSIX link, with no normalisation, and a `CanonicalRelPosixLink` which is a `RelPosixLink` with normalisation applied. I will leave some comments here regarding the result.

Reply:
Okay, after looking at your code I suppose I now understand where my initial intention (including both fields into `RelPosixLink` and use that as a key) would cause problems.

I guess what you got is really the only sensible way.